{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# A primer for PRiMeR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "from scipy.stats import spearmanr\n",
    "from pandas_plink import read_plink\n",
    "\n",
    "import torch\n",
    "from sklearn.impute import SimpleImputer\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.model_selection import train_test_split"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "%autoreload 2\n",
    "from primer.preprocess import gaussianize\n",
    "from primer.gwas import GWAS\n",
    "from primer.mrbased import UVMRbased, MVMRbased\n",
    "from primer.model import PRiMeR\n",
    "from primer.lrm import LRM"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# set path to data\n",
    "risk_factor_file = \"../data/risk_factors.parquet.gzip\"\n",
    "genotype_file = \"../data/genotypes\"\n",
    "covariate_file = \"../data/covariates.parquet.gzip\"\n",
    "sumstats_file = \"../data/target_disease.sumstat\"\n",
    "\n",
    "# followup outcome\n",
    "# this is not required to train PRiMeR, may remain unobserved in applications\n",
    "# here the simulated values are available and are used for validation\n",
    "outcome_file = \"../data/dis_outcome.parquet.gzip\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Risk factor shape: (50000, 30)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Risk_factor_0</th>\n",
       "      <th>Risk_factor_1</th>\n",
       "      <th>Risk_factor_2</th>\n",
       "      <th>Risk_factor_3</th>\n",
       "      <th>Risk_factor_4</th>\n",
       "      <th>Risk_factor_5</th>\n",
       "      <th>Risk_factor_6</th>\n",
       "      <th>Risk_factor_7</th>\n",
       "      <th>Risk_factor_8</th>\n",
       "      <th>Risk_factor_9</th>\n",
       "      <th>...</th>\n",
       "      <th>Risk_factor_20</th>\n",
       "      <th>Risk_factor_21</th>\n",
       "      <th>Risk_factor_22</th>\n",
       "      <th>Risk_factor_23</th>\n",
       "      <th>Risk_factor_24</th>\n",
       "      <th>Risk_factor_25</th>\n",
       "      <th>Risk_factor_26</th>\n",
       "      <th>Risk_factor_27</th>\n",
       "      <th>Risk_factor_28</th>\n",
       "      <th>Risk_factor_29</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.720357</td>\n",
       "      <td>2.063063</td>\n",
       "      <td>-0.380008</td>\n",
       "      <td>0.717654</td>\n",
       "      <td>0.586087</td>\n",
       "      <td>1.183065</td>\n",
       "      <td>0.651498</td>\n",
       "      <td>-0.527197</td>\n",
       "      <td>-0.477083</td>\n",
       "      <td>1.944030</td>\n",
       "      <td>...</td>\n",
       "      <td>0.466053</td>\n",
       "      <td>-0.715269</td>\n",
       "      <td>-0.463753</td>\n",
       "      <td>-1.154618</td>\n",
       "      <td>0.356706</td>\n",
       "      <td>-0.255902</td>\n",
       "      <td>0.662552</td>\n",
       "      <td>-0.943494</td>\n",
       "      <td>1.277293</td>\n",
       "      <td>1.459237</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1.085589</td>\n",
       "      <td>-2.059031</td>\n",
       "      <td>-0.481375</td>\n",
       "      <td>-1.907853</td>\n",
       "      <td>0.401829</td>\n",
       "      <td>-0.702551</td>\n",
       "      <td>0.821321</td>\n",
       "      <td>-2.809480</td>\n",
       "      <td>-0.748050</td>\n",
       "      <td>-1.270696</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.760603</td>\n",
       "      <td>-0.015225</td>\n",
       "      <td>-0.069053</td>\n",
       "      <td>0.169353</td>\n",
       "      <td>1.168955</td>\n",
       "      <td>0.647636</td>\n",
       "      <td>0.471785</td>\n",
       "      <td>0.150448</td>\n",
       "      <td>-0.297059</td>\n",
       "      <td>1.343909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>-0.510135</td>\n",
       "      <td>0.594982</td>\n",
       "      <td>0.015300</td>\n",
       "      <td>0.207112</td>\n",
       "      <td>0.647155</td>\n",
       "      <td>0.102647</td>\n",
       "      <td>0.241790</td>\n",
       "      <td>-0.656938</td>\n",
       "      <td>0.331739</td>\n",
       "      <td>0.816445</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.639291</td>\n",
       "      <td>0.292549</td>\n",
       "      <td>0.763102</td>\n",
       "      <td>0.185799</td>\n",
       "      <td>0.650815</td>\n",
       "      <td>0.009687</td>\n",
       "      <td>0.556082</td>\n",
       "      <td>0.800248</td>\n",
       "      <td>0.208552</td>\n",
       "      <td>0.214982</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>-0.097017</td>\n",
       "      <td>-1.014121</td>\n",
       "      <td>-0.168655</td>\n",
       "      <td>0.623520</td>\n",
       "      <td>0.317475</td>\n",
       "      <td>-0.429457</td>\n",
       "      <td>1.546509</td>\n",
       "      <td>0.169287</td>\n",
       "      <td>0.709109</td>\n",
       "      <td>-0.760504</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.100758</td>\n",
       "      <td>0.772184</td>\n",
       "      <td>-1.351660</td>\n",
       "      <td>-2.096998</td>\n",
       "      <td>-1.091636</td>\n",
       "      <td>-0.308198</td>\n",
       "      <td>0.125971</td>\n",
       "      <td>-1.554938</td>\n",
       "      <td>0.404935</td>\n",
       "      <td>0.665339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>-0.165761</td>\n",
       "      <td>-0.141431</td>\n",
       "      <td>-0.768626</td>\n",
       "      <td>1.090526</td>\n",
       "      <td>-0.957056</td>\n",
       "      <td>-0.059247</td>\n",
       "      <td>-1.342707</td>\n",
       "      <td>-0.839267</td>\n",
       "      <td>1.941332</td>\n",
       "      <td>-0.184904</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.524927</td>\n",
       "      <td>-1.116354</td>\n",
       "      <td>1.551500</td>\n",
       "      <td>-0.301459</td>\n",
       "      <td>-0.729179</td>\n",
       "      <td>-1.059694</td>\n",
       "      <td>-1.624618</td>\n",
       "      <td>-1.407638</td>\n",
       "      <td>-0.166433</td>\n",
       "      <td>-0.567031</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Risk_factor_0  Risk_factor_1  Risk_factor_2  Risk_factor_3  Risk_factor_4  \\\n",
       "0       0.720357       2.063063      -0.380008       0.717654       0.586087   \n",
       "1       1.085589      -2.059031      -0.481375      -1.907853       0.401829   \n",
       "2      -0.510135       0.594982       0.015300       0.207112       0.647155   \n",
       "3      -0.097017      -1.014121      -0.168655       0.623520       0.317475   \n",
       "4      -0.165761      -0.141431      -0.768626       1.090526      -0.957056   \n",
       "\n",
       "   Risk_factor_5  Risk_factor_6  Risk_factor_7  Risk_factor_8  Risk_factor_9  \\\n",
       "0       1.183065       0.651498      -0.527197      -0.477083       1.944030   \n",
       "1      -0.702551       0.821321      -2.809480      -0.748050      -1.270696   \n",
       "2       0.102647       0.241790      -0.656938       0.331739       0.816445   \n",
       "3      -0.429457       1.546509       0.169287       0.709109      -0.760504   \n",
       "4      -0.059247      -1.342707      -0.839267       1.941332      -0.184904   \n",
       "\n",
       "   ...  Risk_factor_20  Risk_factor_21  Risk_factor_22  Risk_factor_23  \\\n",
       "0  ...        0.466053       -0.715269       -0.463753       -1.154618   \n",
       "1  ...       -1.760603       -0.015225       -0.069053        0.169353   \n",
       "2  ...       -1.639291        0.292549        0.763102        0.185799   \n",
       "3  ...       -0.100758        0.772184       -1.351660       -2.096998   \n",
       "4  ...       -0.524927       -1.116354        1.551500       -0.301459   \n",
       "\n",
       "   Risk_factor_24  Risk_factor_25  Risk_factor_26  Risk_factor_27  \\\n",
       "0        0.356706       -0.255902        0.662552       -0.943494   \n",
       "1        1.168955        0.647636        0.471785        0.150448   \n",
       "2        0.650815        0.009687        0.556082        0.800248   \n",
       "3       -1.091636       -0.308198        0.125971       -1.554938   \n",
       "4       -0.729179       -1.059694       -1.624618       -1.407638   \n",
       "\n",
       "   Risk_factor_28  Risk_factor_29  \n",
       "0        1.277293        1.459237  \n",
       "1       -0.297059        1.343909  \n",
       "2        0.208552        0.214982  \n",
       "3        0.404935        0.665339  \n",
       "4       -0.166433       -0.567031  \n",
       "\n",
       "[5 rows x 30 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# risk factor matrix\n",
    "dfRF = pd.read_parquet(risk_factor_file)\n",
    "print('Risk factor shape:', dfRF.shape)\n",
    "dfRF.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Covariate shape: (50000, 2)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sex</th>\n",
       "      <th>age</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.0</td>\n",
       "      <td>53.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>49.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1.0</td>\n",
       "      <td>69.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.0</td>\n",
       "      <td>58.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.0</td>\n",
       "      <td>77.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   sex   age\n",
       "0  0.0  53.0\n",
       "1  1.0  49.0\n",
       "2  1.0  69.0\n",
       "3  0.0  58.0\n",
       "4  0.0  77.0"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# covariates matrix\n",
    "dfC = pd.read_parquet(covariate_file)\n",
    "print('Covariate shape:', dfC.shape)\n",
    "dfC.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Mapping files:   0%|                                                                                                                            | 0/3 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_2186740/1726984066.py:8: FutureWarning: The 'delim_whitespace' keyword in pd.read_csv is deprecated and will be removed in a future version. Use ``sep='\\s+'`` instead\n",
      "  bim, fam, bed = read_plink(str(genotype_file))\n",
      "Mapping files:  33%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–‹                                                                             | 1/3 [00:00<00:00,  4.70it/s]/tmp/ipykernel_2186740/1726984066.py:8: FutureWarning: The 'delim_whitespace' keyword in pd.read_csv is deprecated and will be removed in a future version. Use ``sep='\\s+'`` instead\n",
      "  bim, fam, bed = read_plink(str(genotype_file))\n",
      "Mapping files: 100%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 3/3 [00:00<00:00,  9.36it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Genotype shape: (50000, 755)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sid1</th>\n",
       "      <th>sid2</th>\n",
       "      <th>sid3</th>\n",
       "      <th>sid4</th>\n",
       "      <th>sid5</th>\n",
       "      <th>sid6</th>\n",
       "      <th>sid7</th>\n",
       "      <th>sid8</th>\n",
       "      <th>sid9</th>\n",
       "      <th>sid10</th>\n",
       "      <th>...</th>\n",
       "      <th>sid746</th>\n",
       "      <th>sid747</th>\n",
       "      <th>sid748</th>\n",
       "      <th>sid749</th>\n",
       "      <th>sid750</th>\n",
       "      <th>sid751</th>\n",
       "      <th>sid752</th>\n",
       "      <th>sid753</th>\n",
       "      <th>sid754</th>\n",
       "      <th>sid755</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 755 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   sid1  sid2  sid3  sid4  sid5  sid6  sid7  sid8  sid9  sid10  ...  sid746  \\\n",
       "0   0.0   0.0   0.0   1.0   0.0   0.0   0.0   0.0   0.0    0.0  ...     0.0   \n",
       "1   0.0   1.0   1.0   0.0   0.0   0.0   0.0   1.0   0.0    1.0  ...     0.0   \n",
       "2   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0    1.0  ...     0.0   \n",
       "3   0.0   0.0   0.0   0.0   0.0   0.0   1.0   0.0   0.0    0.0  ...     0.0   \n",
       "4   1.0   1.0   0.0   0.0   0.0   0.0   0.0   0.0   1.0    1.0  ...     1.0   \n",
       "\n",
       "   sid747  sid748  sid749  sid750  sid751  sid752  sid753  sid754  sid755  \n",
       "0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0  \n",
       "1     0.0     0.0     0.0     0.0     0.0     0.0     1.0     0.0     0.0  \n",
       "2     0.0     0.0     0.0     0.0     0.0     1.0     0.0     0.0     0.0  \n",
       "3     0.0     0.0     0.0     0.0     0.0     1.0     0.0     0.0     0.0  \n",
       "4     0.0     1.0     0.0     1.0     0.0     1.0     1.0     0.0     0.0  \n",
       "\n",
       "[5 rows x 755 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# To obtain the genetic variants:\n",
    "# 1. Run GWAS on every trait in the risk factor cohort\n",
    "# 2. Combine all the GWAS results into a single joint summary statistics file\n",
    "# 3. Harmonize the joint summary statistics with our outcome sumstats\n",
    "# 4. Perform LD clumping to obtain a set of independent genetic variants and save ad plink binary file\n",
    "# 5. Use pandas_plink to read plink binary file and convert to pandas Dataframe\n",
    "\n",
    "bim, fam, bed = read_plink(str(genotype_file))\n",
    "dfG = 2 - pd.DataFrame(bed.compute().T, columns=bim[\"snp\"].values) # pandas_plink reads 2 as reference allele => convert to minor allele count\n",
    "print(\"Genotype shape:\", dfG.shape)\n",
    "dfG.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sumstat shape: (755, 3)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>beta</th>\n",
       "      <th>beta_ste</th>\n",
       "      <th>pv</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>sid1</th>\n",
       "      <td>0.009139</td>\n",
       "      <td>0.011262</td>\n",
       "      <td>0.417067</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sid2</th>\n",
       "      <td>-0.004640</td>\n",
       "      <td>0.008037</td>\n",
       "      <td>0.563756</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sid3</th>\n",
       "      <td>0.001874</td>\n",
       "      <td>0.008754</td>\n",
       "      <td>0.830482</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sid4</th>\n",
       "      <td>0.022847</td>\n",
       "      <td>0.009466</td>\n",
       "      <td>0.015790</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sid5</th>\n",
       "      <td>-0.001009</td>\n",
       "      <td>0.014778</td>\n",
       "      <td>0.945590</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          beta  beta_ste        pv\n",
       "sid1  0.009139  0.011262  0.417067\n",
       "sid2 -0.004640  0.008037  0.563756\n",
       "sid3  0.001874  0.008754  0.830482\n",
       "sid4  0.022847  0.009466  0.015790\n",
       "sid5 -0.001009  0.014778  0.945590"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# \n",
    "dfss_outcome = pd.read_csv(sumstats_file, sep=\"\\t\", index_col=0)\n",
    "print('Sumstat shape:', dfss_outcome.shape)\n",
    "dfss_outcome.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Followup outcome shape: (50000, 1)\n",
      "Note: this is not required to train PRiMeR, may remain unobserved in applications. Here the simulated values are available and are used for validation.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.046186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>-0.503982</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>-1.643379</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>-0.474728</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>-0.424113</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          0\n",
       "0  0.046186\n",
       "1 -0.503982\n",
       "2 -1.643379\n",
       "3 -0.474728\n",
       "4 -0.424113"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# followup outcome\n",
    "dfo = pd.read_parquet(outcome_file)\n",
    "print('Followup outcome shape:', dfo.shape)\n",
    "print('Note: this is not required to train PRiMeR, may remain unobserved in applications. Here the simulated values are available and are used for validation.')\n",
    "dfo.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# check indices match\n",
    "assert (dfC.index==dfRF.index).all(), 'outch'\n",
    "assert (dfC.index==dfo.index).all(), 'outch'\n",
    "assert (dfC.index==dfG.index).all(), 'outch'\n",
    "assert (dfss_outcome.index==dfG.columns).all(), 'outch'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Preprocess Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Impute and gaussianize Risk Factors\n",
    "dfRF = pd.DataFrame(\n",
    "    gaussianize(SimpleImputer(strategy=\"mean\").fit_transform(dfRF)),\n",
    "    index=dfRF.index,\n",
    "    columns=dfRF.columns,\n",
    ")\n",
    "\n",
    "# Imputing Genos\n",
    "dfG = pd.DataFrame(\n",
    "    SimpleImputer(strategy=\"mean\").fit_transform(dfG),\n",
    "    index=dfG.index,\n",
    "    columns=dfG.columns,\n",
    ")\n",
    "\n",
    "# Standardize covariates and add intercept\n",
    "dfC[dfC.columns] = StandardScaler().fit_transform(dfC)\n",
    "dfC['intercept'] = 1\n",
    "\n",
    "# Filter rare variants\n",
    "s_before = dfG.shape[1]\n",
    "af = 0.5 * dfG.mean(0)\n",
    "maf = np.minimum(af, 1 - af)\n",
    "Ikeep = list(maf >= 0.01)\n",
    "dfG = dfG.loc[:, Ikeep]\n",
    "s_after = dfG.shape[1]\n",
    "dfss_outcome = dfss_outcome.loc[Ikeep, :]\n",
    "\n",
    "# Converting to float64\n",
    "dfRF = dfRF.astype(np.float64)\n",
    "dfG = dfG.astype(np.float64)\n",
    "dfC = dfC.astype(np.float64)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Split in train and test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# define train / val splits\n",
    "idxs_train, idxs_val = train_test_split(\n",
    "    np.arange(dfRF.shape[0]), test_size=0.2, random_state=42, shuffle=True\n",
    ")\n",
    "\n",
    "# train: used to train PRiMeR\n",
    "dfRF_train = dfRF.iloc[idxs_train].copy()\n",
    "dfG_train = dfG.iloc[idxs_train].copy()\n",
    "dfC_train = dfC.iloc[idxs_train].copy()\n",
    "\n",
    "# train outcome: only used to train the LRM\n",
    "dfo_train = dfo.iloc[idxs_train].copy()\n",
    "\n",
    "# val: used to validate predictor vs followup outcome\n",
    "dfRF_val = dfRF.iloc[idxs_val].copy()\n",
    "dfo_val = dfo.iloc[idxs_val].copy()\n",
    "\n",
    "# standardize the  riskfactors on the training set\n",
    "scaler = StandardScaler()\n",
    "dfRF_train[dfRF_train.columns] = scaler.fit_transform(dfRF_train)\n",
    "dfRF_val[dfRF_val.columns] = scaler.transform(dfRF_val)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run the models"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### MR based models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# compute gwas summary stats for risk factors (for mr based models)\n",
    "gwas_rf = GWAS(dfRF_train.values, dfC_train.values)\n",
    "gwas_rf.process(dfG_train.values)\n",
    "beta_rf = gwas_rf.getBetaSNP()\n",
    "pval_rf = gwas_rf.getPv()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Prediction model based on univate Mendelian randomization\n",
    "uvmr = UVMRbased(pv_iv=5e-8, min_ivs=5, fwer_exp=0.05)\n",
    "dfres_uvmr = uvmr.train(\n",
    "    beta_rf,\n",
    "    pval_rf,\n",
    "    dfss_outcome[\"beta\"].values,\n",
    "    dfss_outcome[\"beta_ste\"].values,\n",
    "    dfRF_train.columns,\n",
    ")\n",
    "\n",
    "pred_uvmr = uvmr.predict(dfRF_val.values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Prediction model based on multivariable Mendelian randomization\n",
    "mvmr = MVMRbased()\n",
    "dfres_mvmr = mvmr.train(\n",
    "    beta_rf,\n",
    "    dfss_outcome[\"beta\"].values,\n",
    "    dfss_outcome[\"beta_ste\"].values,\n",
    "    dfRF_train.columns,\n",
    ")\n",
    "\n",
    "pred_mvmr = mvmr.predict(dfRF_val.values)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Probabilistic Risk Modeling using Mendelian Randomization (PRiMeR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "RF = torch.tensor(dfRF_train.values, dtype=torch.float32)\n",
    "G = torch.tensor(dfG_train.values, dtype=torch.float32)\n",
    "F = torch.tensor(dfC_train.values, dtype=torch.float32)\n",
    "\n",
    "beta_o = torch.tensor(dfss_outcome[\"beta\"].values, dtype=torch.float32).reshape(-1, 1)\n",
    "beta_se_o = torch.tensor(dfss_outcome[\"beta_ste\"].values, dtype=torch.float32).reshape(\n",
    "    -1, 1\n",
    ")\n",
    "\n",
    "E_v = torch.tensor(dfRF_val.values, dtype=torch.float32)\n",
    "o_v = torch.tensor(dfo_val.values, dtype=torch.float32)\n",
    "# RF.shape, G.shape, F.shape, beta_o.shape, beta_se_o.shape, E_t.shape, o_t.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# fit primer lin\n",
    "primer_lin = PRiMeR(RF, G, beta_o, beta_se_o, F=F, lin=True)\n",
    "primer_lin.optimize_lbfgs()\n",
    "primerlin_pred = primer_lin.predict(E_v).detach().numpy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 1000/1000 [00:34<00:00, 28.64it/s]\n"
     ]
    }
   ],
   "source": [
    "# full nonlinear primer\n",
    "primer = PRiMeR(RF, G, beta_o, beta_se_o, F=F, lin=False)\n",
    "primer.log_ve.data[:] = primer_lin.log_ve.data[:]\n",
    "primer.log_vn.data[:] = primer_lin.log_vn.data[:]\n",
    "dmr_history = primer.optimize_sgd(\n",
    "    epochs=1000, lr=1e-2, E=E_v, e_real=o_v, rocauc=False\n",
    ")\n",
    "primer_pred = primer.predict(E_v).detach().numpy()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Model with same functional form as PRiMeR but with access to followup data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 1000/1000 [00:10<00:00, 95.11it/s]\n"
     ]
    }
   ],
   "source": [
    "o_train = torch.tensor(dfo_train.values, dtype=torch.float32)\n",
    "lrm = LRM(RF, o_train, \"elu\", \"regression\")\n",
    "lrm_history = lrm.optimize_sgd(epochs=1000, lr=1e-2, E_val=None, o_val=None)\n",
    "lrm_pred = lrm.predict(E_v).detach().numpy()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Compute metrics and print table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>UVMR-based</th>\n",
       "      <td>0.662906</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MVMR-based</th>\n",
       "      <td>0.701448</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PRiMeR-LIN</th>\n",
       "      <td>0.701476</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PRiMeR</th>\n",
       "      <td>0.731347</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LRM (gold standard)</th>\n",
       "      <td>0.818394</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            0\n",
       "UVMR-based           0.662906\n",
       "MVMR-based           0.701448\n",
       "PRiMeR-LIN           0.701476\n",
       "PRiMeR               0.731347\n",
       "LRM (gold standard)  0.818394"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_results = pd.DataFrame(\n",
    "    data=np.concatenate(\n",
    "        [pred_uvmr, pred_mvmr, primerlin_pred, primer_pred, lrm_pred], axis=1\n",
    "    ),\n",
    "    columns=[\"UVMR-based\", \"MVMR-based\", \"PRiMeR-LIN\", \"PRiMeR\", \"LRM (gold standard)\"],\n",
    ")\n",
    "\n",
    "df_results.apply(lambda x: spearmanr(x, dfo_val.values)[0]).to_frame()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
